types/underscore/underscore-tests.ts

```diff
@@ -12,13 +12,14 @@ _.map({ one: 1, two: 2, three: 3 }, (value, key) => value * 3);
 let plucked: string[] = _.map([{key: 'apples'}, {key: 'oranges'}], 'key');
 
 //var sum = _.reduce([1, 2, 3], (memo, num) => memo + num, 0);    // https://typescript.codeplex.com/workitem/1960
-var sum = _.reduce<number, number>([1, 2, 3], (memo, num) => memo + num, 0);
-sum = _.reduce<number, number>([1, 2, 3], (memo, num) => memo + num); // memo is optional #issue 5 github
-sum = _.reduce<string, number>({'a':'1', 'b':'2', 'c':'3'}, (memo, numstr) => memo + (+numstr));
+_.reduce([1, 2, 3], (memo, num) => memo + num); // $ExpectType number
+_.reduce([1, 2, 3], (memo, num) => memo + num, 0); // $ExpectType number
```

Comment:
I ran into an interesting issue with the `UnderscoreStatic` overloads of these functions specifically, which is why there are separate list and dictionary overloads of them at this time. For some reason if I use a `Collection` constraint and a `MemoCollectionIterator` in `UnderscoreStatic`, a `memo` value of `0` is not widened to `number` but instead causes the result type to be `0`, which doesn't work well. Super surprisingly, though, that behavior does not also occur in `Underscore` and `_Chain` overloads. I should probably look for an issue in the TS repository that explains why this might be occurring and file an issue asking about it if I can't find one.

Reply:
Oh yeah, this issue also only happens when the type for `memo` is not explicitly specified in the iteratee, which is why my tests didn't catch it since I always declare the iteratee separately with explicit parameters. The alternation between a partial and full iterator signature was something I added as an attempt to figure out what was going on here, and I decided to keep it because it seemed like a nice case to cover.

Reply:
I think the example below that shows the difference in type inference given a function vs. class generic `V` is as simple as I can make it and still have it repro:

```TS
interface MemoIterator<TResult> {
    (memo: TResult, num: number): TResult;
}

type MemoCollectionIterator<TResult, V> =
    V extends number[] ? MemoIterator<TResult>
    : never;

interface FunctionGeneric {
    reduce<V, TResult>(
        collection: V,
        iteratee: MemoCollectionIterator<TResult, V>,
        memo: TResult
    ): TResult;
}

interface ClassGeneric<V> {
    reduce<TResult>(
        collection: V,
        iteratee: MemoCollectionIterator<TResult, V>,
        memo: TResult
    ): TResult;
}

declare const functionGeneric: FunctionGeneric;
declare const classGeneric: ClassGeneric<number[]>;

functionGeneric.reduce([1, 2, 3], (memo, num) => memo + num, 0); // TResult is inferred as 0
classGeneric.reduce([1, 2, 3], (memo, num) => memo + num, 0); // TResult is inferred as number
```

Reply:
I guess `V` isn't being inferred in the `ClassGeneric` case, so it probably isn't adding much, which I guess indicates that it's the inference of `V` that's causing issues.

```TS
interface MemoIterator<TResult> {
    (memo: TResult, num: number): TResult;
}

type MemoCollectionIterator<TResult, V> =
    V extends number[] ? MemoIterator<TResult>
    : never;

interface Test {
    reduce<V, TResult>(
        collection: V,
        iteratee: MemoCollectionIterator<TResult, V>,
        memo: TResult
    ): TResult;

    reduce2<TResult>(
        collection: number[],
        iteratee: MemoCollectionIterator<TResult, number[]>,
        memo: TResult
    ): TResult;
}

declare const test: Test;

test.reduce([1, 2, 3], (memo, num) => memo + num, 0); // TResult is inferred as 0
test.reduce2([1, 2, 3], (memo, num) => memo + num, 0); // TResult is inferred as number
```

Reply:
Even in the state machine case, I still think the use of a type union is correct because one still needs to figure out what one currently has and what one ends up with at the end of the call (though I suppose one could still possibly infer that without doing type checks by doing current index and collection size checks instead). I agree, though, that that is a good reason to still allow the memo to be augmented to more than just the collection element type, and I appreciate you raising that point. I feel like it would probably still be worth forcing the result type to at least include the collection element type when no memo is provided, though. I think that could probably be managed by updating the overload to include something like `TResult extends TypeOfCollection<V> = TypeOfCollection<V>`, but I'll experiment a bit and see whether that constraint works correctly with union types.

Reply:
Hmm, darn, another case that I didn't try that I should have is giving `reduce` and `reduceRight` an empty collection. It looks like these functions return `undefined` in the case of empty collections, so results should probably also be updated to include that.

Reply:
I think I have a good set of overloads in theory, but I'm still working on getting them to be happy in practice since there's still some oddness going on with constant types not widening and also now a compilation issue that only occurs in TS3.3 and below. Here's what I'm thinking would be ideal to allow for a customizable final result but ensure that either the type of the memo (if one is specified) or the type of the collection (if one is not) is included in the results.

```TS
        reduce<V extends Collection<any>, TResult = TypeOfCollection<V>>(
            collection: V,
            iteratee: MemoCollectionIterator<TypeOfCollection<V>, TResult | TypeOfCollection<V>, V>,
            memo?: undefined,
            context?: any
        ): TResult | TypeOfCollection<V> | undefined;
        reduce<V extends Collection<any>, TMemo, TResult = TMemo>(
            collection: V,
            iteratee: MemoCollectionIterator<TypeOfCollection<V>, TResult | TMemo, V>,
            memo: TMemo,
            context?: any
        ): TResult | TMemo | undefined;
```

Reply:
Alright, aside from the widening issue the first signature was working, it was just the second one that was having issues and I was finally able to resolve those in a way that seems satisfactory by dropping `TMemo` in favor of just using `TResult` for `memo` (which should be fine since it will still take any value that is of any one of the types in the result type union). I've added tests that demonstrate union type memo iterators working which includes putting back the test I removed; let me know what you think @jgonggrijp. With regards to the test that I added back, I put in the full expected result type but it's worth noting that it still forces the collection type to be included if the collection type is omitted as shown below.

```TS
_.reduce<_.Dictionary<string>, number>({ 'a': '1', 'b': '2', 'c': '3' }, (memo, numstr) => (+memo) + (+numstr)); // $ExpectType string | number | undefined
```